Add editor block inserted analytic event

dcalhoun opened the pull request:
## Related PRs

* https://github.com/WordPress/gutenberg/pull/35272
* https://github.com/wordpress-mobile/gutenberg-mobile/pull/4066
* https://github.com/wordpress-mobile/WordPress-Android/pull/15406

## Description

* Fixes https://github.com/wordpress-mobile/gutenberg-mobile/issues/4052.
* Fixes https://github.com/wordpress-mobile/gutenberg-mobile/issues/4027. 

Analytic events originating from JavaScript are communicated in the form of a raw string event name. This adds logic to track events by converting a finite list of event names to `WPAnalyticEvent` assignments.

## Testing

Relevant event properties:
- `block_name`
- `blog_id`
- `inner_block`
- `parent_block_name`

### Block Inserter
1. Launch the block editor.
2. Insert a block.
3. ℹ️ **Expected:** `wpios_editor_block_inserted` event arrives with the relevant properties. 

### Copy/Paste
1. Insert a Columns block. 
2. Populate blocks within each column, an Image in one and a Paragraph in the other. 
3. Tap the three-dot settings button on the Columns block. 
4. Tap "Copy block." 
5. Tap the three-dot settings button on the Columns block. 
6. Tap "Paste block."
7. ℹ️ **Expected:** A `wpios_editor_block_inserted` event arrives for each of the blocks (Columns, Column (2), Image, Paragraph) with relevant properties. 

### Rich Text Split
1. Launch the block editor.
2. Insert a Pull Quote block. 
3. Place the cursor at the end of the citation input. 
4. Press the <kbd>Return</kbd> key. 
5. ℹ️ **Expected:** `wpios_editor_block_inserted` event arrives with the relevant properties. 

## Regression Notes
1. Potential unintended areas of impact
    Existing analytics event could malfunction. 
2. What I did to test those areas of impact (or what existing automated tests I relied on)
    Manually verified a random sampling of analytic events fired as expected. 
3. What automated tests I added (or what prevented me from doing so)
    None. There are no preexisting tests for the modified code/file. The lack of familiarity with iOS testing inhibited me from writing the foundational test structure for `WPAnalyticsEvent`. Open to adding the tests with some help. 

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

dcalhoun committed:
Add editor block inserted analytic event

Analytic events originating from JavaScript are communicated in the form
of a raw string event name. This adds logic to track events by
converting a finite list of event names to `WPAnalyticEvent`
assignments.

dcalhoun commented on WordPress/Classes/Utility/Analytics/WPAnalyticsEvent.swift:
```diff
@@ -631,4 +634,25 @@ extension WPAnalytics {
+        props[WPAppAnalyticsKeyBlogID] = blog.dotComID
+        props[WPAppAnalyticsKeySiteType] = blog.isWPForTeams() ? WPAppAnalyticsValueSiteTypeP2 : WPAppAnalyticsValueSiteTypeBlog
+
+        var event: WPAnalyticsEvent?
+        switch eventName {
+        case "editor_block_inserted": event = .gutenbergEditorBlockInserted
+        default: event = nil
+        }
```
This switch statement manages the event name string to `WPAnalyticEvent` conversion. It is currently placed here in an attempt to keep all analytic type-related logic within this one file.

dcalhoun replied:
The sibling https://github.com/wordpress-mobile/WordPress-Android/pull/15406 follows this approach as well, as the required enum is not available within the Gutenberg-related fragment. So, it likely makes sense to retain this approach for WPiOS.

dcalhoun commented on WordPress/Classes/ViewRelated/Gutenberg/GutenbergViewController.swift:
```diff
@@ -1014,6 +1014,10 @@ extension GutenbergViewController: GutenbergBridgeDelegate {
         gutenbergSettings.blockTypeImpressions = impressions
     }
+
+    func gutenbergDidRequestSendEventToHost(_ eventName: String, properties: [AnyHashable: Any]) -> Void {
+        WPAnalytics.trackBlockEditorEvent(eventName, properties: properties, blog: post.blog)
+    }
```
This function is another candidate location for the event name string to `WPAnalyticEvent` conversion. The argument for this location would be an attempt to avoid "leaking" JavaScript-related patterns into the WPiOS app, i.e. keep the idea of string event names out of the `WPAnalyticsEvent` file.

dcalhoun committed:
Merge branch 'develop' of github.com:wordpress-mobile/WordPress-iOS into feature/block-editor-analytics

dcalhoun committed:
Merge branch 'develop' of github.com:wordpress-mobile/WordPress-iOS into feature/block-editor-analytics

dcalhoun committed:
Update Gutenberg ref

mchowning commented on WordPress/Classes/Utility/Analytics/WPAnalyticsEvent.swift:
```diff
@@ -669,4 +672,25 @@ extension WPAnalytics {
+        var event: WPAnalyticsEvent?
+        switch eventName {
+        case "editor_block_inserted": event = .gutenbergEditorBlockInserted
+        default: event = nil
```
Would it be useful to log something here if we hit the default case since, if I understand correctly, that should never happen unless we've implemented something incorrectly? This is certainly not a blocker.

dcalhoun replied:
Solid improvement. Addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/17243/commits/ffb71878bf875701e5b47015f9728add3900183e.

mchowning commented on WordPress/Classes/Utility/Analytics/WPAnalyticsEvent.swift:
```diff
@@ -669,4 +672,25 @@ extension WPAnalytics {
+
+        if event != nil {
+             WPAnalytics.track(event!, properties: props)
+        }
```
Could we avoid updating the props with the blog info here and instead use track(eventName, properties, blog) to take care of them?

```suggestion
    static func trackBlockEditorEvent(_ eventName: String, properties: [AnyHashable: Any], blog: Blog) {

        var event: WPAnalyticsEvent?
        switch eventName {
        case "editor_block_inserted": event = .gutenbergEditorBlockInserted
        default: event = nil
        }

        if event != nil {
             WPAnalytics.track(event!, properties: props, block: blog)
        }
```

dcalhoun replied:
Definitely a good idea. Addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/17243/commits/a26a8e7a6c3b8b9c516a30c579c5ba4764410fda.

mchowning commented on WordPress/Classes/Utility/Analytics/WPAnalyticsEvent.swift:
```diff
@@ -669,4 +672,25 @@ extension WPAnalytics {
 
+    /// This will call each registered tracker and fire the given event.
+    /// - Parameters:
+    ///   - eventName: a `String` that represents the event name
```
Just a thought. I don't feel strongly.

```suggestion
    ///   - eventName: a `String` that represents the Block Editor event name
```

mchowning reviewed (commented):
Nice work @dcalhoun ! Tested this and it is working well. I left a few comments, but they're generally pretty minor issues.

dcalhoun committed:
Leverage existing code

Avoid repeating the generation of site-specific event properties by
leveraging the existing method signature available.

dcalhoun committed:
Add nuance to block editor analytics documentation

Clarify the specificity of block editor event names.

Co-authored-by: Matt Chowning <matt.chowning@automattic.com>

dcalhoun committed:
Log warning when a Block Editor event is ignored

Warn developers when the event passed for analytics tracking is ignored
due to it not matching a defined `WPAnalyticEvent` case.

mchowning reviewed (commented):
Changes look good, and retested to verify everything is still working great. :shipit:

dcalhoun committed:
Update Gutenberg ref

dcalhoun committed:
Merge branch 'develop' of github.com:wordpress-mobile/WordPress-iOS into feature/block-editor-analytics
